MuxExoPlayer/src/main/java/com/mux/stats/sdk/muxstats/MuxBaseExoPlayer.java

```diff
@@ -779,251 +787,214 @@ public void outputLog(String tag, String msg) {
-      BandwidthMetricData loadData = new BandwidthMetricData();
-      loadData.setRequestError(e.toString());
-      if (dataSpec != null && dataSpec.uri != null) {
-        loadData.setRequestUrl(dataSpec.uri.toString());
-        loadData.setRequestHostName(dataSpec.uri.getHost());
-      }
-      switch (dataType) {
-        case C.DATA_TYPE_MANIFEST:
-          loadData.setRequestType("manifest");
-          break;
-        case C.DATA_TYPE_MEDIA:
-          loadData.setRequestType("media");
-          break;
-        default:
-          return null;
-      }
-      loadData.setRequestErrorCode(null);
-      loadData.setRequestErrorText(e.getMessage());
-      return loadData;
-    }
+    TrackGroupArray availableTracks;
+    HashMap<String, BandwidthMetricData> loadedSegments = new HashMap<>();
 
-    public BandwidthMetricData onLoadCanceled(DataSpec dataSpec) {
-      BandwidthMetricData loadData = new BandwidthMetricData();
-      loadData.setRequestCancel("genericLoadCanceled");
-      if (dataSpec != null && dataSpec.uri != null) {
-        loadData.setRequestUrl(dataSpec.uri.toString());
-        loadData.setRequestHostName(dataSpec.uri.getHost());
+    public BandwidthMetricData onLoadError(String segmentUrl,
+        IOException e) {
+      BandwidthMetricData segmentData = loadedSegments.get(segmentUrl);
+      if (segmentData == null) {
+        segmentData = new BandwidthMetricData();
+        // TODO We should see how to put minimal stats here !!!
       }
-      loadData.setRequestType("media");
-      return loadData;
+      segmentData.setRequestError(e.toString());
+      // TODO see what error codes are
+      segmentData.setRequestErrorCode(-1);
+      segmentData.setRequestErrorText(e.getMessage());
+      segmentData.setRequestResponseEnd(System.currentTimeMillis());
+      return segmentData;
+    }
+
+    public BandwidthMetricData onLoadCanceled(String segmentUrl) {
+      BandwidthMetricData segmentData = loadedSegments.get(segmentUrl);
+      if (segmentData == null) {
+        segmentData = new BandwidthMetricData();
+        // TODO We should see how to put minimal stats here !!!
+      }
+      segmentData.setRequestCancel("genericLoadCanceled");
+      segmentData.setRequestResponseEnd(System.currentTimeMillis());
+      return segmentData;
     }
 
-    protected BandwidthMetricData onLoad(DataSpec dataSpec, int dataType,
-        Format trackFormat, long mediaStartTimeMs, long mediaEndTimeMs,
-        long elapsedRealtimeMs, long loadDurationMs, long bytesLoaded) {
-      BandwidthMetricData loadData = new BandwidthMetricData();
-      if (bytesLoaded > 0) {
-        loadData.setRequestBytesLoaded(bytesLoaded);
-      }
+    protected BandwidthMetricData onLoad(long mediaStartTimeMs, long mediaEndTimeMs,
+        String segmentUrl, int dataType, String host
+    ) {
+      BandwidthMetricData segmentData = new BandwidthMetricData();
+      // TODO I have no idea how to get RequestSttart timestamp
```

Comment:
What do you plan to do in this case?

Reply:
So far nothing,  there is no way to get this data, I just noted that it is needed, maybe in future ExoPlayer will allow us to collect this.

Reply:
Please rephrase the comment, something like:
```suggestion
      // TODO RequestStart timestamp is currently not available from ExoPlayer
```